### PR TITLE
Redesign for Functions Hosting config

### LINKED ISSUE #8923: Functions hosting configuration redesign
**Current implementation has the downsides:**
1.	The configuration updates dynamically during application lifetime, this can cause issues if a feature flag is using in different places of the code base. 
2.	The configuration updates on `GetValue` call and caches for 10 minutes, so the update operation blocks on the config file read and parse.

```
    public interface IFunctionsHostingConfiguration
    {
        bool FunctionsWorkerDynamicConcurrencyEnabled { get; }

        string GetValue(string key, string defaultValue = null);
    }
```

**Proposed implementation.**

1.	Using` IOptions<FunctionsHostingConfig>` for static configuration:
```
private IOptions<FunctionsHostingConfig> _functionsHostingConfig;
public SampleHostingService(IOptions<FunctionsHostingConfig> functionsHostingConfig)
{
      _functionsHostingConfig = functionsHostingConfig;
}
..
bool enabled = _functionsHostingConfig.Value.SomeFeatureEnabled;
```

2.	Using `IOptionsMonitor<FunctionsHostingConfig>` for dynamic configuration:
```
private readonly IDisposable _configChangeHandle;
private IOptionsMonitor<FunctionsHostingConfig> _monitor;

public SampleHostingService(IOptions<FunctionsHostingConfig> functionsHostingConfig, IOptionsMonitor<FunctionsHostingConfig> monitor)
{
_monitor = monitor;
_configChangeHandle = monitor.OnChange(UpdateConfiguration);
}

private void UpdateConfiguration(FunctionsHostingConfig config)
{
            bool enabled = config.Value.SomeFeatureEnabled;
}
Or
bool enabled = _monitor.CurrentValue.SomeFeatureEnabled;
```
3.	Passing the hosting config to language worker processes as env variables.

### PR DESCRIPTION
**Pull request by alrod**
### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-functions-host/issues/8923

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

To add a flag to config:
image

Note that the config is passed to a language workers as on the language workers creation/specialization. The config would not be updated in the language worker if it was modified during function host life time.

### CODE REVIEW

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptionsSetup.cs
Comment (jviau) [inline src/WebJobs.Script/Config/FunctionsHostingConfigOptionsSetup.cs]: Is there a reason we are reaching into this specific provider? If possible, I would prefer to see options not have intimate knowledge of the providers. Instead, I recommend having a specific config section we use, such as `_configuration.GetSection("FunctionsPlatform")` (or a better section name if you have one).

File: src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs
Comment (jviau) [inline src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs:13]: naming nit: I like the naming of _Platform_ over _Hosting_ here - as the file you are loading is the "FunctionsPlatformConfig" (by the environment variable).

File: src/WebJobs.Script/Config/FunctionsHostingConfigProvider.cs
Comment (jviau) [inline src/WebJobs.Script/Config/FunctionsHostingConfigProvider.cs:37]: Maybe in the future we can have a structured config file, like `ini` or `json`.

In the meantime, I recommend prefixing all the keys loaded with a well-known configuration section, so that options configuration can rely on section and not specific provider. A prefix like:

``` CSharp
.ToDictionary(a => $"FunctionsPlatform:{a[0]}", a => a[1], StringComparer.OrdinalIgnoreCase);
```

File: src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs
Comment (jviau) [inline src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs]: Is the `ConfigurationChangeTokenSource` necessary here? I thought that config and options reload will all be built into the file watcher from `FunctionsHostingConfigSource`.

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs:26]: can you add feature flag for target based scaling as well?

**Comment by pragnagopa** [comment]
@alrod - please update the PR description design details and how this can be used to turn on /off features in Functions Host and Language Workers.

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs
Comment (alrod) [reply src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs:26]: For TBS it's not make sense to add the properties as the flags are dynamic (assembly names). First step is getting assembly name where a target scalers is defined and then checking the config if the flag exists.

File: src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs
Comment (madelinegordon) [reply src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs:13]: Is the feature in Antares called "hosting config"? /wondering if that is the reason for the name?

File: src/WebJobs.Script/Config/FunctionsHostingConfigProvider.cs
Comment (madelinegordon) [inline src/WebJobs.Script/Config/FunctionsHostingConfigProvider.cs]: nit : comment "A file-based hosting configuration"

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptionsSetup.cs
Comment (alrod) [reply src/WebJobs.Script/Config/FunctionsHostingConfigOptionsSetup.cs]: The section is a good idea.

File: src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs
Comment (alrod) [reply src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs:13]: It's because the config is stored in "hosting" SQL database on stamp level. This name was already discussed previously on initial feature commit.

File: src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs
Comment (alrod) [reply src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs]: This was necessary to fire `OnChange `for `IOptionsMonitor `and after introducing the section it fires without the line - removed.

**Comment by alrod** [comment]
> @alrod - please update the PR description design details and how this can be used to turn on /off features in Functions Host and Language Workers.

Added the information to the PR description.

File: src/WebJobs.Script/ScriptConstants.cs
Comment (jviau) [inline src/WebJobs.Script/ScriptConstants.cs]: Think you have an extra whitespace here:
```suggestion
        public static readonly string FunctionsHostingConfigSectionName = "FunctionsHostingConfig";
```

File: src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs
Comment (jviau) [inline src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs]: Do you need both `IOptions` and `IOptionsMonitor`? You can call `IOptionsMonitor.CurrentValue` to get the equivalent of `IOptions.Value`.

**Review by jviau — CHANGES_REQUESTED** [review]

File: test/WebJobs.Script.Tests/Configuration/FunctionsHostingConfigOptionsSetupTest.cs
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Configuration/FunctionsHostingConfigOptionsSetupTest.cs:22]: add test for delimeter `;`

File: test/WebJobs.Script.Tests/Configuration/FunctionsHostingConfigProviderTest.cs
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Configuration/FunctionsHostingConfigProviderTest.cs]: same here - add test for delimeter `;`

File: src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs:24]: please add comments here - that indicate this path is for windows SKUs only.

File: src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Config/FunctionsHostingConfigSource.cs:20]: please add comments here on the possible values of the file path when running on Linux SKUs

**Review by pragnagopa — COMMENTED** [review]
Added minor comments. LGTM - Approving. Please wait for final approval from Functions Runtime Team - cc @fabiocav

**Comment by pragnagopa** [comment]
cc @CooperLink - FYI - to rebase https://github.com/Azure/azure-functions-host/pull/8891 on top of changes in this PR.

File: src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs
Comment (alrod) [reply src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs]: Good point!

**Review by jviau — APPROVED** [review]

File: src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs
Comment (madelinegordon) [inline src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs]: nit - `_hostingConfigOnChange` (missing the g in hosting)

**Review by madelinegordon — APPROVED** [review]

**Pull request merged**
